http/vibe/http/fileserver.d

```diff
@@ -427,3 +419,254 @@ private void sendFileImpl(scope HTTPServerRequest req, scope HTTPServerResponse
+				res.statusCode = HTTPStatus.notModified;
+			else
+				res.statusCode = HTTPStatus.preconditionFailed;
+			res.writeVoidBody();
+			return true;
+		}
+	}
+
+	if (last_modified != SysTime.init) {
+		// https://tools.ietf.org/html/rfc7232#section-3.3
+		string ifModifiedSince = req.headers.get("If-Modified-Since");
+		if (ifModifiedSince.length) {
+			const check = lastModifiedString == ifModifiedSince ||
+				last_modified <= parseRFC822DateTimeString(ifModifiedSince);
+			if (check) {
+				res.statusCode = HTTPStatus.notModified;
+				res.writeVoidBody();
+				return true;
+			}
+		}
+
+		// https://tools.ietf.org/html/rfc7232#section-3.4
+		string ifUnmodifiedSince = req.headers.get("If-Unmodified-Since");
+		if (ifUnmodifiedSince.length) {
+			const check = lastModifiedString != ifUnmodifiedSince
+				|| last_modified > parseRFC822DateTimeString(ifUnmodifiedSince);
+			if (check) {
+				res.statusCode = HTTPStatus.preconditionFailed;
+				res.writeVoidBody();
+				return true;
+			}
+		}
+	}
+
+	return false;
+}
+
+/**
+	Represents an Entity-Tag value for use inside HTTP Cache headers.
+
+	Standards: https://tools.ietf.org/html/rfc7232#section-2.3
+*/
+struct ETag
+{
+	bool weak;
+	string tag;
+
+	static ETag parse(string s)
+	{
+		enforce!ConvException(s.endsWith('"'));
+
+		if (s.startsWith(`W/"`)) {
+			ETag ret = { weak: true, tag: s[3 .. $ - 1] };
+			return ret;
+		} else if (s.startsWith('"')) {
+			ETag ret;
+			ret.tag = s[1 .. $ - 1];
+			return ret;
+		} else {
+			throw new ConvException(`ETag didn't start with W/" nor with " !`);
+		}
+	}
+
+	string toString() const @property
+	{
+		return text(weak ? `W/"` : `"`, tag, '"');
+	}
+
+	/**
+		Encodes the bytes with URL Base64 to a human readable string and returns an ETag struct wrapping it.
+	 */
+	static ETag fromBytesBase64URLNoPadding(scope const(ubyte)[] bytes, bool weak = false)
+	{
+		import std.base64 : Base64URLNoPadding;
+
+		return ETag(weak, Base64URLNoPadding.encode(bytes).idup);
+	}
+
+	/**
+		Hashes the input bytes with md5 and returns an URL Base64 encoded representation as ETag.
+	 */
+	static ETag md5(T...)(bool weak, T data)
+	{
+		import std.digest.md : md5Of;
+
+		return fromBytesBase64URLNoPadding(md5Of(data), weak);
+	}
+}
+
+/**
+	Matches a given match expression with a specific ETag. Can allow or disallow weak ETags and supports multiple tags.
+
+	Standards: https://tools.ietf.org/html/rfc7232#section-2.3.2
+*/
+bool cacheMatch(string match, ETag etag, bool allow_weak)
+{
+	if (match == "*") {
+		return true;
+	}
+
+	if ((etag.weak && !allow_weak) || !match.length) {
+		return false;
+	}
+
+	auto allBytes = match.representation;
+
+	for (auto range = allBytes; !range.empty; range.popFront)
+	{
+		auto start = range = range.find('"').drop(1);
+		range = range.find('"');
+		if (range.empty)
+			break;
+
+		const check = start[0 .. &range[0] - &start[0]];
+		const checkWeak = match[0 .. &start[0] - &allBytes[0]].endsWith(`W/"`);
```

Comment:
Can't you do sth. simpler like e.g. 

```d
auto range = match.representation;
bool isWeak = range.skipOver("W/");
// if weak -> must check whether allowed (weak comparison)
if (isWeak && !allow_weak) {
  return false;
```

And now checking for the etag becomes simply:

```
// etag must start/end with "
if (range.length < 2 || range[0] != '"' || range[$ - 1] != '"')
    return false;

return range[1 .. $ -1] == etag.tag;
```

Or if you need to check for multiple etags, with a loop of course.